### ISSUE TITLE
Run integration tests in GitHub Actions

### ISSUE THREAD

**Issue by oguzhanunlu** [snowplow/snowplow-android-tracker#328]
The next release, `1.3.0`, is expected to include a new test suite for the tracker, integration tests. These tests use Snowplow Micro, to send some events and retrieve the validation results per Micro API specs.

The under-development test suite can be run as expected locally while running them in Travis may end up failed. Sometimes they work as expected and I suspect this is because we didn't fix the emulator launched inside Travis. We need to fix this issue with running ui tests in Travis environment. Until then, we'll run ui tests by hand locally during development.

**Comment by plastiv** [comment]
As you have pointed out having code that is running inside of the android app (AndroidJunitTest) to communicate to the local server running outside of the android on host system (snowplow micro) is a major headache. 

Things I noticed, while trying to make this integration:
* Discoverability. `localhost` inside of the android resolves to different address, so there is a requirement to use external service from outside network
* Docker daemon requirement. To start snowplow micro docker daemon is needed, which not every cloud/ci provider easily gets access to due to exposing vulnerabilities. Check https://github.com/GoogleContainerTools/kaniko for more info
* More importantly, docker daemon requirement means, that we can't control, start/stop snowplow micro from the test itself. Meaning, snowplow micro becomes a static resource, with all of the downsides of having static state in tests -> each test has to additionally validate/clear state, **tests can't be run in parallel** 

Was it considered to restructure test approach to either:

* Make snowplow micro running in android jvm. So, similarly, to https://github.com/square/okhttp/tree/master/mockwebserver it would be possible to have everything running inside of the android and control it from android junit instrumentation test
* Or vice versa, remove hard dependency of android snowplow sdk on android by making it pluggable. We can manage snow-plow-micro on jvm unit tests with https://github.com/testcontainers/testcontainers-java . But after that snowplow-android-tracker is not ready to ignore missing AdvertisementID and so on, even with robolectric (android runtime on jvm)

**Comment by oguzhanunlu** [comment]
Hi @plastiv , thanks for the detailed comment. Your observations are correct. We can always question and reconsider our approaches including how we test.

My opinions on your recommendations are below.

The reason Micro has an HTTP interface is to enable more usage scenarios compared to targeting only one platform, apart from other snowplow ecosystem related reasons. There might be a future version of Micro supporting gRPC communication to handle Micro operations directly in JVM but server would stay there at the end of the day, i.e. no parallel testing in that case too

I checked `testcontainers` project and it seems useful but Travis CI offers docker support and I'm not sure if it is worth the effort to switch how we change docker container. I couldn't see how it'll improve our process.

I think that no matter how we launch/run Micro, as long as we have one Micro instance running, we won't run tests in parallel. In case we consider launching multiple Micro instances, we can do that in the current setup too.

Please let me know if I'm missing anything.

**Comment by matus-tomlein** [comment]
Closing since we don't have Micro tests implemented currently – that will be the next action in this regard but we should have a separate issue for it.
